python.d/tomcat.chart.py

```diff
@@ -16,61 +18,141 @@
+        'lines': [
+            ["eden_used", 'used', "absolute", 1, 1048576],
+            ["eden_commited", 'commited', "absolute", 1, 1048576],
+            ["eden_max", 'max', "absolute", 1, 1048576]
+        ]},
+    'jvm_survivor': {
+        'options': [None, "Survivor Memory Usage", "MB", "memory", "tomcat.jvm_survivor", "area"],
+        'lines': [
+            ["survivor_used", 'used', "absolute", 1, 1048576],
+            ["survivor_commited", 'commited', "absolute", 1, 1048576],
+            ["survivor_max", 'max', "absolute", 1, 1048576]
+        ]},
+    'jvm_tenured': {
+        'options': [None, "Tenured Memory Usage", "MB", "memory", "tomcat.jvm_tenured", "area"],
+        'lines': [
+            ["tenured_used", 'used', "absolute", 1, 1048576],
+            ["tenured_commited", 'commited', "absolute", 1, 1048576],
+            ["tenured_max", 'max', "absolute", 1, 1048576]
+        ]},
 }
 
-
 class Service(UrlService):
     def __init__(self, configuration=None, name=None):
         UrlService.__init__(self, configuration=configuration, name=name)
         self.url = self.configuration.get('url', "http://127.0.0.1:8080/manager/status?XML=true")
+        self.connector_name = self.configuration.get('connector_name', None)
         self.order = ORDER
         self.definitions = CHARTS
 
     def check(self):
-        netloc = urlparse(self.url).netloc.rpartition(':')
-        if netloc[1] == ':': port = netloc[2]
-        else: port = 80
-        
-        self.regex_jvm = compile(r'<jvm>.*?</jvm>')
-        self.regex_connector = compile(r'[a-z-]+%s.*?/connector' % port)
-        self.regex = compile(r'([\w]+)=\\?[\'\"](\d+)\\?[\'\"]')
-        
-        return UrlService.check(self)
+        return UrlService.check(self) and self._get_data() != None
 
     def _get_data(self):
         """
         Format data received from http request
         :return: dict
         """
-        data = self._get_raw_data()
-        if data:
-            jvm = self.regex_jvm.findall(data) or ['']
-            connector = self.regex_connector.findall(data) or ['']
-            data = dict(self.regex.findall(''.join([jvm[0], connector[0]])))
-        
-        return data or None
+        data = None
+        raw_data = self._get_raw_data()
+        if raw_data:
+            xml = ET.fromstring(raw_data)
+            data = {}
+
+            jvm = xml.find('jvm')
+
+            connector = None
+            if self.connector_name:
+                for conn in xml.findall('connector'):
+                    if conn.get('name') == self.connector_name:
+                        connector = conn
+                        break
+            else:
+                connector = xml.find('connector')
+
+            memory = jvm.find('memory')
+            data['free']  = memory.get('free')
+            data['total'] = memory.get('total')
+            data['used']  = int(data['total']) - int(data['free'])
```

Comment:
Not sure about this. Dict `get` returns the result or `None`. In case of `None` `int(None)` will raise the Exception.